### PR TITLE
advert creation and edititing

### PR DESCRIPTION
**Pull request by volmelnyk**
pull created

### CODE REVIEW

File: Adesk.BLL/DTOs/AddAdvertDTO.cs [hunk 1]
```diff
@@ -12,22 +12,15 @@ namespace Adesk.BLL.DTOs
...
-        [Required(ErrorMessage = ErrorMessages.NoDescription)]
-        [MinLength(1)]
-        [MaxLength(4000, ErrorMessage = ErrorMessages.LongDescription)]
         public string Description { get; set; }
 
-        [Required(ErrorMessage = ErrorMessages.NoSubCathegory)]
+
         public long SubCathegoryId { get; set; }
 
         public IList<SelectListItem> Cathegories { get; set; }
```
Comment (wolf8196) [inline Adesk.BLL/DTOs/AddAdvertDTO.cs]: outdated.needs to be removed

File: Adesk.BLL/DTOs/EditAdvertDTO.cs [hunk 2]
```diff
@@ -14,23 +14,14 @@ public class EditAdvertDTO
         [Required]
```
Comment (wolf8196) [inline Adesk.BLL/DTOs/EditAdvertDTO.cs]: all dto attributes need to be removed

File: Adesk.BLL/DTOs/EditAdvertDTO.cs [hunk 3]
```diff
@@ -14,23 +14,14 @@ public class EditAdvertDTO
...
 
-        [Required(ErrorMessage = ErrorMessages.NoDescription)]
-        [MinLength(1)]
-        [MaxLength(4000, ErrorMessage = ErrorMessages.LongDescription)]
         public string Description { get; set; }
 
-        [Required(ErrorMessage = ErrorMessages.NoSubCathegory)]
         public long SubCathegoryId { get; set; }
 
         public IList<SelectListItem> Cathegories { get; set; }
```
Comment (wolf8196) [inline Adesk.BLL/DTOs/EditAdvertDTO.cs]: outdated. same as above

Comment (wolf8196) [reply Adesk.BLL/DTOs/EditAdvertDTO.cs]: we can use one dto for both editing and adding, by making Id property null, when adding

Comment (wolf8196) [reply Adesk.BLL/DTOs/EditAdvertDTO.cs]: attributes can be left, but without error strings. we need to add them to other dtos used for posting data, and add small check for modelstate as it is now in auth controller

File: Adesk.BLL/Services/AdvertService.cs [hunk 4]
```diff
@@ -134,5 +134,77 @@ public IEnumerable<AdvertDTO> GetUserAdverts(string id)
...
+            {
+                var advert = x.AdvertRepository.GetById(id);
+                var editEdvert = new EditAdvertDTO
+                {
+                    Id = id,
+                    Title = advert.Title,
+                    Price = advert.Price,
+                    Description = advert.Description,
+                    Photo = advert.Photo,
+                    Cathegories = (from c in x.CathegoryRepository.Query()
```
Comment (wolf8196) [inline Adesk.BLL/Services/AdvertService.cs]: this is the ugly stuff that was needed for mvc. needs to be removed

File: Adesk.BLL/Services/AdvertService.cs [hunk 5]
```diff
@@ -134,5 +134,77 @@ public IEnumerable<AdvertDTO> GetUserAdverts(string id)
...
+                                   select new SelectListItem
+                                   {
+                                       Text = c.Name,
+                                       Value = c.Id.ToString(),
+                                       Selected = (advert.SubCathegory.CathegoryId == c.Id) ? true : false
+                                   }).ToList(),
+                    SubCathegories = (from sc in x.SubCathegoryRepository.Query()
```
Comment (wolf8196) [inline Adesk.BLL/Services/AdvertService.cs]: same as above

File: Adesk.BLL/Services/AdvertService.cs [hunk 6]
```diff
@@ -134,5 +134,77 @@ public IEnumerable<AdvertDTO> GetUserAdverts(string id)
...
+                advert.Price = newAdvert.Price;
+                advert.SubCathegoryId = newAdvert.SubCathegoryId;
+                advert.Photo = newAdvert.Photo ?? AdeskConstants.DefaultAdvertPhoto;
+
+                x.AdvertRepository.Update(advert);
+                x.Save();
+            });
+        }
+
+        public void AddAdvert(AddAdvertDTO newAdvert, string userId)
```
Comment (wolf8196) [inline Adesk.BLL/Services/AdvertService.cs]: probably need to return Id of inserted advert to have the ability to navigate user to its details page

File: Adesk.Web/Controllers/WebApiControllers/AdvertsController.cs [hunk 7]
```diff
@@ -47,6 +48,14 @@ public IHttpActionResult DeleteOwnerAdvert(long id)
             return Ok();
         }
 
+        [HttpPost]
+        [Authorize(Roles = "User")]
+        public IHttpActionResult AddOwnerAdvert(AddAdvertDTO newAdvert, string userId)
```
Comment (wolf8196) [inline Adesk.Web/Controllers/WebApiControllers/AdvertsController.cs]: need to take the userId from token

File: Adesk.sln [hunk 8]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio 15
-VisualStudioVersion = 15.0.26730.12
+VisualStudioVersion = 15.0.26730.15
```
Comment (wolf8196) [inline Adesk.sln]: looks like you forgot to add dto files to solution. it won't build on other teammembers' machines

File: Adesk.BLL/Services/AdvertService.cs [hunk 9]
```diff
@@ -134,5 +134,77 @@ public IEnumerable<AdvertDTO> GetUserAdverts(string id)
                         .ToList();
             });
         }
+
+        public EditAdvertDTO GetEditAdvert(long id)
```
Comment (wolf8196) [inline Adesk.BLL/Services/AdvertService.cs]: needs better naming

Comment (wolf8196) [reply Adesk.BLL/Services/AdvertService.cs]: this method is not in the interface. you wont be able to access it from controller

File: Adesk.Web/Controllers/WebApiControllers/AdvertsController.cs [hunk 10]
```diff
@@ -47,6 +49,43 @@ public IHttpActionResult DeleteOwnerAdvert(long id)
             return Ok();
         }
 
+        [HttpPost]
+        [Authorize(Roles = "User")]
+        [Route("add")]
```
Comment (rvovk) [inline Adesk.Web/Controllers/WebApiControllers/AdvertsController.cs]: no need to name this. see below

File: Adesk.BLL/DTOs/AddEditAdvertDTO.cs [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+using System.Web.Mvc;
+using Adesk.Common;
+
+namespace Adesk.BLL.DTOs
+{
+    public class AddEditAdvertDTO
+    {
+        public long Id { get; set; }
+        [Required]
+        public string Title { get; set; }
```
Comment (wolf8196) [inline Adesk.BLL/DTOs/AddEditAdvertDTO.cs]: add max length

File: Adesk.BLL/DTOs/AddEditAdvertDTO.cs [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+        [Required]
+        public decimal Price { get; set; }
+        [Required]
+        public string Photo { get; set; }
```
Comment (wolf8196) [inline Adesk.BLL/DTOs/AddEditAdvertDTO.cs]: not required. if there is no phot. we add the default one in service

File: Adesk.BLL/DTOs/AddEditAdvertDTO.cs [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+        [Required]
+        public string Description { get; set; }
```
Comment (wolf8196) [inline Adesk.BLL/DTOs/AddEditAdvertDTO.cs]: this is not required. it should just have max length

File: Adesk.Web/Controllers/WebApiControllers/AdvertsController.cs (hunk 10 above)
Comment (wolf8196) [inline Adesk.Web/Controllers/WebApiControllers/AdvertsController.cs]: there is not need to name this route. see comment below

File: Adesk.Web/Controllers/WebApiControllers/AdvertsController.cs [hunk 14]
```diff
@@ -47,6 +49,43 @@ public IHttpActionResult DeleteOwnerAdvert(long id)
...
+        [Route("edit/{id}")]
+        public IHttpActionResult GetEditAdvert(long id)
+        {
+            return Json(advertService.GetEditAdvert(id));
+        }
+
+        [HttpPost]
+        [Authorize(Roles = "User")]
+        [Route("edit")]
+        public IHttpActionResult EditAdvert(AddEditAdvertDTO editAdvert)
```
Comment (wolf8196) [inline Adesk.Web/Controllers/WebApiControllers/AdvertsController.cs]: updates should be made with put request. in that case you won't need to name the route

**Pull request merged**